### PR TITLE
use the *adjusted* callee type in effect checking

### LINKED ISSUE #28776: effect checking should really be using `expr_ty_adjusted`
Reproduced on: ```rustc 1.5.0-nightly (6108e8c3e 2015-09-28)```

```Rust
unsafe fn blow_up() {
    *(0xcccccccc as *mut u8) = 0;
}

fn main() {
    (&blow_up)();
}
```

### PR DESCRIPTION
**Pull request by arielb1**
Fixes #28776

r? @eddyb

### CODE REVIEW

File: src/librustc/lib.rs [hunk 1]
```diff
@@ -105,6 +105,7 @@ pub mod front {
 }
 
 pub mod middle {
+    pub mod expr_use_visitor; // STAGE0: increase glitch immunity
```
Comment (eddyb) [inline src/librustc/lib.rs]: This is starting to get a *little* bit silly :smiley:...

**Comment by brson** [comment]
What is the impact of this change? Please provide more context.

**Comment by arielb1** [comment]
@brson 

Previous versions of rustc could allow calling an unsafe function outside of an unsafe block under conditions involving autoderef.

**Comment by brson** [comment]
@arielb1 Thanks!

**Pull request merged**